pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -60,196 +56,151 @@
-      _postingListLengthIntBuffer.put(dictId, _postingListLengthIntBuffer.get(dictId) + 1);
+      putInt(_valueBuffer, _nextValueId++, dictId);
+      putInt(_postingListIndexBuffer, dictId, getInt(_postingListIndexBuffer, dictId) + 1);
     }
+    putInt(_numValuesBuffer, _nextDocId++, length);
   }
 
   @Override
   public void seal() throws IOException {
-    // Construct the posting list
-
-    // Create posting list value buffer
-    _postingListValueByteBuffer =
-        MmapUtils.allocateDirectByteBuffer(_numValues * V1Constants.Numbers.INTEGER_SIZE, null,
-            "bitmap posting list value buffer for: " + _columnName);
-    IntBuffer postingListValueIntBuffer = _postingListValueByteBuffer.asIntBuffer();
-
-    // Create posting list current offset buffer and initialize it using posting list length buffer
-    _postingListCurrentOffsetByteBuffer =
-        MmapUtils.allocateDirectByteBuffer(_cardinality * V1Constants.Numbers.INTEGER_SIZE, null,
-            "bitmap posting list current offset buffer for: " + _columnName);
-    IntBuffer postingListCurrentOffsetIntBuffer = _postingListCurrentOffsetByteBuffer.asIntBuffer();
-    int postingListStartOffset = 0;
+    // Calculate posting list index for each dictionary Id
+    int postingListIndex = 0;
     for (int dictId = 0; dictId < _cardinality; dictId++) {
-      postingListCurrentOffsetIntBuffer.put(dictId, postingListStartOffset);
-      postingListStartOffset += _postingListLengthIntBuffer.get(dictId);
+      int length = getInt(_postingListIndexBuffer, dictId);
+      putInt(_postingListIndexBuffer, dictId, postingListIndex);
+      postingListIndex += length;
     }
 
-    // Dump values into posting list
+    // Construct posting list
+    _postingListValueBuffer = new LBuffer((long) _numValues * Integer.BYTES);
     if (_singleValue) {
       for (int docId = 0; docId < _numDocs; docId++) {
-        int dictId = _valueIntBuffer.get(docId);
-        int offset = postingListCurrentOffsetIntBuffer.get(dictId);
-        postingListValueIntBuffer.put(offset++, docId);
-        postingListCurrentOffsetIntBuffer.put(dictId, offset);
+        int dictId = getInt(_valueBuffer, docId);
+        int index = getInt(_postingListIndexBuffer, dictId);
+        putInt(_postingListValueBuffer, index, docId);
+        putInt(_postingListIndexBuffer, dictId, index + 1);
       }
     } else {
-      int valueOffset = 0;
+      int valueId = 0;
       for (int docId = 0; docId < _numDocs; docId++) {
-        int numValues = _numValuesIntBuffer.get(docId);
+        int numValues = getInt(_numValuesBuffer, docId);
         for (int i = 0; i < numValues; i++) {
-          int dictId = _valueIntBuffer.get(valueOffset++);
-          int offset = postingListCurrentOffsetIntBuffer.get(dictId);
-          postingListValueIntBuffer.put(offset++, docId);
-          postingListCurrentOffsetIntBuffer.put(dictId, offset);
+          int dictId = getInt(_valueBuffer, valueId++);
```

Comment:
Lines 128-131 and 138-141 look similar. Anychance they can be a single method call?

Reply:
the dictId are fetched from different buffer, so I don't think we need a single method call for 3 lines of same code